File: 09_DataStructure_Algorithm/0518_day51/03_피로도/피로도.java
class Solution {
    int answer;
    
    public int solution(int k, int[][] dungeons) {
        answer = -1;
        
        dfs(k, dungeons, new boolean[dungeons.length], 0);
        
        return answer;
    }
    
    void dfs(int k, int[][] dungeons, boolean[] visited, int count) {
        answer = Math.max(count, answer);
        
        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                dfs(k - dungeons[i][1], dungeons, visited, count + 1);
                visited[i] = false;
            }
        }
    }
}

File: 09_DataStructure_Algorithm/0518_day51/03_피로도/피로도.py
def solution(k, dungeons):
    answer = -1
    
    n = len(dungeons)
    visited = [False] * n
    
    def permutations(count, k):
        nonlocal answer
        answer = max(answer, count)
        
        for idx in range(n):
            if not visited[idx] and k >= dungeons[idx][0]:
                visited[idx] = True
                permutations(count + 1, k - dungeons[idx][1])
                visited[idx] = False
    
    permutations(0, k)
        
    return answer